src/test/ui/did_you_mean/issue-40396.stderr

```diff
@@ -80,17 +80,13 @@ error[E0615]: attempted to take value of method `collect` on type `std::ops::Ran
   --> $DIR/issue-40396.rs:2:13
    |
 LL |     (0..13).collect<Vec<i32>>();
-   |             ^^^^^^^
-   |
-   = help: maybe a `()` to call it is missing?
+   |             ^^^^^^^ help: use parentheses to call the method: `collect()`
 
 error[E0615]: attempted to take value of method `collect` on type `std::ops::Range<{integer}>`
   --> $DIR/issue-40396.rs:18:13
    |
 LL |     (0..13).collect<Vec<i32>();
-   |             ^^^^^^^
-   |
-   = help: maybe a `()` to call it is missing?
+   |             ^^^^^^^ help: use parentheses to call the method: `collect()`
```

Comment:
> The suggestions in `src/test/ui/did_you_mean/issue-40396.stderr` are suboptimal. I could check if the containing expression is `BinOp`, but I'm not sure if that's general enough. Any ideas?

Though this is a case that will need to be accounted for specifically, as the ideal case would be for the parser to actually be [able to parse type setting without the turbofish](https://github.com/rust-lang/rust/pull/53578) and emit a customized relevant suggestion to use the turbofish. If we do that, then this error will not trigger.

Reply:
What would be the "perfect" suggestion in that case? IMO, that's borderline "unrecoverable": the code is attempting (assumption 1) to set a type argument without using turbofish and (assumption 2) calling that method. Because this requires not one but two assumptions, the likelihood of a suggestion being incorrect increases. In this case we would emit the following, which I think is "good enough" (at least for now):

```
error: chained comparison operators require parentheses
  --> src/main.rs:11:12
   |
11 |     Foo.bar<foo>();
   |            ^^^^^^
   |
   = help: use `::<...>` instead of `<...>` if you meant to specify type arguments
   = help: or use `(...)` if you meant to specify fn arguments

error[E0615]: attempted to take value of method `bar` on type `Foo`
  --> src/main.rs:11:9
   |
11 |     Foo.bar<foo>();
   |         ^^^ help: use parentheses to call the method: `bar()`

error[E0308]: mismatched types
  --> src/main.rs:11:17
   |
11 |     Foo.bar<foo>();
   |                 ^^ expected bool, found ()
   |
   = note: expected type `bool`
              found type `()`
```

Reply:
I suppose a better example is this:

```rust
struct Foo;

impl Foo {
    fn bar(&self) -> i32 {
        0
    }
}

fn main() {
    let Vec = 1;
    Foo.bar < Vec;
}
```

The user is attempting to compare the result of a method call, but forgot the parens. If we don't emit the suggestion because it's parsed as part of a `BinOp` `<`, it would be incorrect. How can I distinguish this from a missing turbofish?

Reply:
We'd need to keep track of either `NodeId`s or `Span`s where the turbofish note due to chained comparison operators in a set owned by the parser. We then need to keep that arround in the `Sess`. When finding this case in the type checker, we look back at the `sess` to see if the place where we're about to suggest adding a `()` is covered by the turbofish suggestion or not. All of this should be done in a separate PR, as it might quickly become very involved.